homeassistant/components/developer_credentials/__init__.py

```diff
@@ -0,0 +1,270 @@
+        id_manager,
+    )
+    storage_listener = DeveloperCredentialsStorageListener(hass)
+    storage_collection.async_add_listener(storage_listener.updated)
+    await storage_collection.async_load()
+
+    collection.StorageCollectionWebsocket(
+        storage_collection, DOMAIN, DOMAIN, CREATE_FIELDS, UPDATE_FIELDS
+    ).async_setup(hass)
+
+    # Allow future registration of local oauth implementations
+    config_entry_oauth2_flow.async_register_local_apis(hass)
+
+    return True
+
+
+# Creates AbstractOAuth2Implementation given a DeveloperCredential for a specific domain
+AuthImplFactory = Callable[
+    [str, DeveloperCredential], config_entry_oauth2_flow.AbstractOAuth2Implementation
+]
+
+
+def _get_auth_domain(domain: str, developer_credential: DeveloperCredential) -> str:
+    """Return the OAuth2 flow implementation domain."""
+    return slugify(f"{domain}.{developer_credential.client_id}")
+
+
+async def _async_register_auth_implementation(
+    hass: HomeAssistant,
+    domain: str,
+    developer_credential: DeveloperCredential,
+) -> None:
+    """Register an OAuth2 flow implementation for an integration."""
+    auth_domain = _get_auth_domain(domain, developer_credential)
+    authorization_server = await _async_get_authorization_server(hass, domain)
+    if not authorization_server:
+        raise ValueError("No authorization server registered for %s" % domain)
+    auth_impl = config_entry_oauth2_flow.LocalOAuth2Implementation(
+        hass,
+        auth_domain,
+        developer_credential.client_id,
+        developer_credential.client_secret,
+        authorization_server.authorize_url,
+        authorization_server.token_url,
+    )
+    unsub = config_entry_oauth2_flow.async_register_implementation(
+        hass, domain, auth_impl
+    )
+    hass.data[DOMAIN][domain][auth_domain] = unsub
+
+
+def _async_unregister_auth_implementation(
+    hass: HomeAssistant, domain: str, developer_credential: DeveloperCredential
+) -> None:
+    """Register an OAuth2 flow implementation for an integration."""
+    auth_domain = _get_auth_domain(domain, developer_credential)
+    unsub = hass.data[DOMAIN][domain][auth_domain]
+    unsub()
+
+
+async def _async_get_authorization_server(
+    hass: HomeAssistant, domain: str
+) -> AuthorizationServer | None:
+    """Return the AuthorizationServer for the integration domain."""
+    if domain not in hass.data[DOMAIN]:
+        return None
+    get_authorization_server = hass.data[DOMAIN][domain][
+        DeveloperCredentialsType.AUTHORIZATION_SERVER.value
+    ]
+    authorization_server = await get_authorization_server(hass)
+    if not authorization_server:
+        return None
+    return authorization_server
+
+
+async def _async_get_developer_credential(
+    hass: HomeAssistant, domain: str
+) -> DeveloperCredential | None:
+    """Return the DeveloperCredential for the integration domain."""
+    if domain not in hass.data[DOMAIN]:
```

Comment:
Since this is a private method, this can't happen ? Or would this method be called by something that is called by other integrations ?

Reply:
Yeah, ok, this method isn't doing anything useful. I'm inlining it where it was called.  (I think this had other uses originally that were removed as i developed it)